### PR TITLE
Feat/ade 495/edit function macro

### PR DESCRIPTION
**Pull request by robinMcA**

### CODE REVIEW

File: src/main/java/au/com/agiledigital/idea_search/helpers/StructureFieldRenderHelper.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
  */
 public class StructureFieldRenderHelper {
 
-  public static String render(StructuredCategory category, String body) {
+  public static String render(StructuredCategory category, String body ) {
```
Comment (shouvSarker) [inline src/main/java/au/com/agiledigital/idea_search/helpers/StructureFieldRenderHelper.java:24]: Extra space after body? Looks like a bunch of other files have it too, is this related to the style guide?

File: src/main/resources/js/updateMacro.js [hunk 2]
```diff
@@ -0,0 +1,338 @@
+// HTML dialog for simple text input
+const simpleTextInput = `
```
Comment (shouvSarker) [inline src/main/resources/js/updateMacro.js:2]: Can the html snippet be generalised? i.e. a function that generates HTML based on the type of input?

File: src/main/resources/js/updateMacro.js [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+
+/**
+ * Handle the update of the user macro
+ *
+ * @param getNewUserArray function that returns an array of user link elements
+ * @param macroNode to be updated
+ * @param macroParams existing on the macro that is being updated
+ */
+const updateUserList = (getNewUserArray, macroNode, macroParams) => {
+  AJS.$("body").append(userInput);
```
Comment (shouvSarker) [inline src/main/resources/js/updateMacro.js:107]: define `AJS.$ = $` and use `$`?

File: src/main/resources/js/updateMacro.js [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+  const getNewUserArray = () =>
+    getStringsFromInput()
+      .map(
+        (text) =>
+          '<a class="confluence-userlink user-mention current-user-mention userlink-0" data-username="' +
+          text +
+          '" href="/confluence/display/~' +
+          text +
+          '" data-linked-resource-id="65588" data-linked-resource-version="1" data-linked-resource-type="userinfo" ' +
+          'data-base-url="http://localhost:1990/confluence" title="" data-user-hover-bound="true">' +
```
Comment (shouvSarker) [inline src/main/resources/js/updateMacro.js:216]: hard coded url?

File: src/main/resources/js/updateMacro.js [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+      updateUserList(getNewUserArray, macroNode, macroParams);
+      return;
+    case "technologies":
+      addTechnologies(
+        originalBody,
+        getStringsFromInput,
+        macroNode,
+        macroParams
+      );
+      return;
```
Comment (shouvSarker) [inline src/main/resources/js/updateMacro.js:287]: No default case?

File: src/main/resources/templates/pageBlueprint.xml [hunk 6]
```diff
@@ -1,33 +1,33 @@
 <ac:structured-macro ac:name="Idea Search Client Side Scripts" />
-<ac:structured-macro ac:name="Idea Structured Field">
+<ac:structured-macro ac:name="idea-structured-field">
```
Comment (shouvSarker) [inline src/main/resources/templates/pageBlueprint.xml:2]: Change in naming convention?

File: src/main/resources/vm/StructuredFieldTechnology.vm [hunk 7]
```diff
@@ -1,6 +1,6 @@
 #* @vtlvariable name="payload" type="java.lang.reflect.Array" *#
 #if($heading)
-    <h2>Technologies</h2>
+<h2>Technologies</h2>
```
Comment (shouvSarker) [inline src/main/resources/vm/StructuredFieldTechnology.vm:3]: Indent after if?

File: src/main/resources/js/updateMacro.js (hunk 3 above)
Comment (robinMcA) [reply src/main/resources/js/updateMacro.js:107]: It was giving me errors about redefining it, and I have to use some other AJS calls so i did it this way.

File: src/main/resources/vm/StructuredFieldTechnology.vm (hunk 7 above)
Comment (robinMcA) [reply src/main/resources/vm/StructuredFieldTechnology.vm:3]: This matches the other vm files.

File: src/main/java/au/com/agiledigital/idea_search/macros/StructuredField.java [hunk 8]
```diff
@@ -38,9 +41,9 @@ public String execute(Map<String, String> map, String s, ConversionContext conve
     StructuredCategory category = StructuredCategory.fromKey(map.get("category"));
 
     /** Replaces all html tags */
-    String stripped = removeTags(s);
+//    String stripped = removeTags(s);
```
Comment (shouvSarker) [inline src/main/java/au/com/agiledigital/idea_search/macros/StructuredField.java]: Remove dead code? Fixed in develop branch so may merge it in?

File: src/main/resources/js/updateMacro.js (hunk 5 above)
Comment (robinMcA) [reply src/main/resources/js/updateMacro.js:287]: yeah, if it is invoked elsewhere a dialogue won't pop up and do odd things to your page.

File: src/main/resources/js/StructuredField.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
+const $ = AJS.$;
+
+AJS.toInit(() => {
+  // AJS.$("#technologies-input").auiSelect2();
```
Comment (shouvSarker) [inline src/main/resources/js/StructuredField.js]: Remove commented code?

File: src/main/resources/js/status-light.js [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+    <footer class="aui-dialog2-footer">
+        <div class="aui-dialog2-footer-actions">
+            <button id="dialog-submit-button" class="aui-button aui-button-primary">Make it so</button>
+        </div>
+    </footer>
+</section>`
+
+
+var updateMacro = function (macroNode, params, newBody) {
+
```
Comment (shouvSarker) [inline src/main/resources/js/status-light.js]: Two blank lines?

File: src/main/resources/js/status-light.js [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+          })
+
+          AJS.Confluence.Binder.autocompleteMultiUser();
+
+          AJS.dialog2("#edit-page-dialog").show();
+
+          AJS.$("#dialog-submit-button").click(function (e) {
+            const newTechArray = AJS.$(".select2-search-choice").toArray().map(item => item.innerText).map(text => "<a class=\"confluence-userlink user-mention current-user-mention userlink-0\" data-username=\"" + text +
+              "\" href=\"/confluence/display/~" + text +
+              "\" data-linked-resource-id=\"65588\" data-linked-resource-version=\"1\" data-linked-resource-type=\"userinfo\" data-base-url=\"http://localhost:1990/confluence\" title=\"\" data-user-hover-bound=\"true\">" + text +
```
Comment (shouvSarker) [inline src/main/resources/js/status-light.js]: Is the `data-linked-resource-id` the same every time?

File: src/main/resources/js/updateMacro.js (hunk 2 above)
Comment (robinMcA) [reply src/main/resources/js/updateMacro.js:2]: I'm not sure if it would be worth it.
There are enough differences between the to make it annoying.

File: src/main/resources/js/updateMacro.js (hunk 3 above)
Comment (shouvSarker) [reply src/main/resources/js/updateMacro.js:107]: Ah, okay. Yeah, I noticed the different types of `AJS.something` calls.

File: src/main/resources/templates/pageBlueprint.xml (hunk 6 above)
Comment (robinMcA) [reply src/main/resources/templates/pageBlueprint.xml:2]: The macro needs to be called by the name, I think this is so that there is no white space in the name.
better than the title case name

File: src/main/resources/js/updateMacro.js [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+   */
+  const getNewUserArray = () =>
+    getStringsFromInput()
+      .map(
+        (text) =>
+          '<a class="confluence-userlink user-mention current-user-mention userlink-0" data-username="' +
+          text +
+          '" href="/confluence/display/~' +
+          text +
+          '" data-linked-resource-id="65588" data-linked-resource-version="1" data-linked-resource-type="userinfo" ' +
```
Comment (shouvSarker) [inline src/main/resources/js/updateMacro.js:215]: hard coded resource id?

Comment (robinMcA) [reply src/main/resources/js/updateMacro.js:215]: looking at this, seems like user links are hard.

File: src/main/resources/js/updateMacro.js [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+      }),
+    },
+  });
+  AJS.dialog2("#edit-page-dialog-tech").show();
+
+  AJS.$("#dialog-submit-button").click((e) => {
+    const oldTechArray = originalBody.split(",");
+
+    // Construct the new technology list
+    const newTechText = [...newTechnologies(), ...oldTechArray]
```
Comment (SamMurphyDev) [inline src/main/resources/js/updateMacro.js:163]: Could a set be used here to handle de-duplicating for you?

File: src/main/resources/js/updateMacro.js [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+      )
+      .join(", ");
+
+  // The category assigned to the macro
+  const categoryArray = macroNode.dataset.macroParameters.split("=");
+  // Put the macro params into the correct format
+  const macroParams = { [categoryArray[0]]: categoryArray[1] };
+
+  const handleFormSubmit = (category, newBodyText) => {
+    const formID = "#edit-page-dialog-" + category;
```
Comment (SamMurphyDev) [inline src/main/resources/js/updateMacro.js:230]: A template literal can be used here

```suggestion
    const formID = `#edit-page-dialog-${category}`;
```

File: src/main/resources/js/updateMacro.js (hunk 13 above)
Comment (shouvSarker) [reply src/main/resources/js/updateMacro.js]: Used a set.

File: src/main/resources/js/updateMacro.js (hunk 14 above)
Comment (shouvSarker) [reply src/main/resources/js/updateMacro.js]: Updated.

File: src/main/resources/js/updateMacro.js (hunk 12 above)
Comment (shouvSarker) [reply src/main/resources/js/updateMacro.js]: As per discussion, decided to drop user link functionality for now.

**Comment by shouvSarker** [comment]
Converted to draft because the following bug gets introduced with current PR. 

Will open it again after that is fixed.
Screenshot from 2020-12-21 11-07-37

**Comment by shouvSarker** [comment]
Converted to draft because the following bug gets introduced with current PR. 

Will open it again after that is fixed.
Screenshot from 2020-12-21 11-07-37

**Comment by shouvSarker** [comment]
Bug fixed and ready for review.

**Review by SamMurphyDev — APPROVED** [review]
